SDL/SDLJoystick.cpp

```diff
@@ -11,167 +11,209 @@ extern "C" {
 	}
 }
 
-SDLJoystick::SDLJoystick(bool init_SDL ): thread(NULL), running(true) {
+SDLJoystick::SDLJoystick(bool init_SDL ): thread(NULL), running(true), player1Controller(NULL), player1JoyInstanceID(-1) {
 	if (init_SDL)
 	{
-		SDL_setenv("SDL_JOYSTICK_ALLOW_BACKGROUND_EVENTS", "1", 0);
-		SDL_Init(SDL_INIT_JOYSTICK | SDL_INIT_VIDEO);
+		SDL_SetHint(SDL_HINT_JOYSTICK_ALLOW_BACKGROUND_EVENTS, "1");
+		SDL_Init(SDL_INIT_JOYSTICK | SDL_INIT_VIDEO | SDL_INIT_GAMECONTROLLER);
 	}
+	
+	if (SDL_GameControllerAddMappingsFromFile("assets/gamecontrollerdb.txt") == -1)
+	{
+			printf("Failed to load control pad mappings, please place gamecontrollerdb.txt in your assets directory\n");
+	}
+	else {
+		printf("Loaded game controller mappings for SDL2\n");
+		setUpPlayer1();
+	}
+}
 
+void SDLJoystick::setUpPlayer1()
+{
 	int numjoys = SDL_NumJoysticks();
-	SDL_JoystickEventState(SDL_ENABLE);
+	int joyIndex = -1;
 	for (int i = 0; i < numjoys; i++) {
-		joys.push_back(SDL_JoystickOpen(i));
-//		printf("Initialized joystick %d: %s",i,SDL_JoystickNameForIndex(i));
-		if(strstr(SDL_JoystickNameForIndex(i),"PLAYSTATION(R)3 Controller"))
-			g_Config.bPS3Controller = true;
+		if (joyIndex == -1 && SDL_IsGameController(i))
+		{
+			joyIndex = i;
+			break;
+		}
+	}
+	if (joyIndex != -1)
+	{
+		printf("Player 1 will be using joystick %d, %s\n", joyIndex, SDL_JoystickNameForIndex(joyIndex));
```

Comment:
_"the PSP only supports one player"_

Correct. Here by "Player 1" I mean player 1 of the PSP, not control pad number 1.

When multiple control pads are connected, if the control pad assigned to player 1 is disconnected, then the new code loops through the remaining control pads connected looking for the first recogonised control pad. If one is found then this control pad then becomes assigned to "player 1" (see the SDL_CONTROLLERDEVICEADDED event handling on line 171).

I was assuming that only control pad should be used to control PPSSPP. If you prefer that PPSSP responds to input from any connected control pad then this can be done but I thought that would be confusing to have more than one control pad acting as "player 1"?